### ISSUE TITLE
File registry is broken and leaks files

### ISSUE THREAD

**Issue by noha** [guillep/pharo-core#14626]
**Bug description**
The way the File registry (FinalizationRegistry) is handling entries is bogus and leaks all files. 

**To Reproduce**
1. file := File openForWriteFileNamed: 'testfile'.
2. file close
3. testfile entry is still in the File registry

**Expected behavior**
After closing the file the entry should have vanished from the file registry

**Version information:**
 - OS:  MacOS but independent 
 - Pharo Version >= 11

**Expected development cost**
Estimated cost is 2-4 hours because the bug is identified just needs a proper solution

**Additional context**
The bug is there because when a file is opened it is added to the File registry with 

```
add: anObject finalizer: finalizer

	self protected: [ 
		ephemeronCollection add:
			(FinalizationRegistryEntry key: anObject value: finalizer container: self) ].
	^ anObject
```

but on closing of the file it does 

```
remove: anEphemeron ifAbsent: aBlock

	^ self protected: [ ephemeronCollection remove: anEphemeron ifAbsent: aBlock ]
```

where anEphemeron is the file stream object itself. Therefor the added object and the to be removed one are not the same therefor it is a miss in the ephemeronCollection and it returns without unregistering the file. I would have provided a fix directly instead of writing this issue but is unclear how in that context a fast removal should work with the wrapped object. So delegating to guille :)

**Comment by Ducasse** [comment]
Hi Norbert 
Are you in P12? Because we got some changes in finalization so this is an important information.

**Comment by jecisc** [comment]
Reading the description of Norbert, the change will not fix this particular issue I think. 
But the report of Norbert is really nice and it should not be that hard to fix the bug :)

**Comment by noha** [comment]
@Ducasse I'm on p11, I did not check p12 just assumed it hasn't changed there. 

I don't get the rationale of having an IdentitySet for the ephemeron management which contains elements that are subclasses of Association where the key and value are both the stream to be managed. But yes everytime I read about ephemerons and understand it, 5 minutes later it is lost :) So I remember something that the association is needed but then it might be possible to use an IdentityDictionary instead of IdentitySet which would allow fast removal

**Comment by guillep** [comment]
Thanks, I'll check it.

In the previous weeks I was working on some improvements on ephemerons, see https://github.com/pharo-project/pharo/pull/14068

Mostly they were performance improvements, but I remember passing through the file tests when doing that.

Now, I did those changes against Pharo12, with the idea to then backport to P11 if OK.
We should check if that PR fixed this issue in P12.

**Comment by noha** [comment]
I tested in p12 and there it works as expected. So this would be worth a backport to fix the p11 release. 

Doing the tests I discovered that there a quite some files in the registry that shouldn't be there. Examples are 

/builds/workspace/uest_and_branch_Pipeline_Pharo12/64/bootstrap-cache/pharo-local/package-cache/System-BasicCommandLineHandler-tonel.1.mcz

and 

OpalCompiler-Core.hermes

So the look like leftovers of the build process. There are 64 of those files in total. Shall I make a ticket for the clean of those files?

**Comment by noha** [comment]
@guillep Any news on this?

**Comment by guillep** [comment]
Yes. Check https://github.com/pharo-project/pharo/issues/14633.
They are actually the same issue. Maybe we should merge them and close the duplicate.

TL;DR; 
 - A fix is coming to the VM
 - then we need to backport the finalization registry fixes to P11

**Comment by guillep** [comment]
With the latest changes, the situation is much better.

I've found one file is still leaking, I need to investigate that.
My current findings are that:
 - the ephemeron is in the registry
 - but it has been marked by the GC as collectable

So it was apparently not finalized.
Maybe there is an answer for that in the build's log.

**Comment by noha** [comment]
thanks for the fix. I'm changing deployment to use it

**Comment by guillep** [comment]
Sure, give us feedback please :)

**Comment by noha** [comment]
Looks good now. I cannot detect any leaks in a server running for multiple days. So I confirm for p11 that it is fixed

**Comment by guillep** [comment]
Thanks @noha for the confirmation!!! I'm glad this goes in the right direction!

I'll close this issue then ^^.
